src/veins/base/connectionManager/ChannelAccess.cc

```diff
@@ -82,56 +82,24 @@ void ChannelAccess::initialize(int stage)
 
 void ChannelAccess::sendToChannel(cPacket* msg)
 {
-    const NicEntry::GateList& gateList = cc->getGateList(getParentModule()->getId());
-    NicEntry::GateList::const_iterator i = gateList.begin();
-
-    if (useSendDirect) {
-        // use Andras stuff
-        if (i != gateList.end()) {
-            simtime_t delay = SIMTIME_ZERO;
-            for (; i != --gateList.end(); ++i) {
-                // calculate delay (Propagation) to this receiving nic
-                delay = calculatePropagationDelay(i->first);
-
-                int radioStart = i->second->getId();
-                int radioEnd = radioStart + i->second->size();
-                for (int g = radioStart; g != radioEnd; ++g) sendDirect(static_cast<cPacket*>(msg->dup()), delay, msg->getDuration(), i->second->getOwnerModule(), g);
-            }
-            // calculate delay (Propagation) to this receiving nic
-            delay = calculatePropagationDelay(i->first);
+    EV_TRACE << "sendToChannel: sending to gates\n";
 
-            int radioStart = i->second->getId();
-            int radioEnd = radioStart + i->second->size();
-            for (int g = radioStart; g != --radioEnd; ++g) sendDirect(static_cast<cPacket*>(msg->dup()), delay, msg->getDuration(), i->second->getOwnerModule(), g);
+    const auto& gateList = cc->getGateList(getParentModule()->getId());
 
-            sendDirect(msg, delay, msg->getDuration(), i->second->getOwnerModule(), radioEnd);
-        }
-        else {
-            EV_WARN << "Nic is not connected to any gates!" << endl;
-            delete msg;
-        }
-    }
-    else {
-        // use our stuff
-        EV_TRACE << "sendToChannel: sending to gates\n";
-        if (i != gateList.end()) {
-            simtime_t delay = SIMTIME_ZERO;
-            for (; i != --gateList.end(); ++i) {
-                // calculate delay (Propagation) to this receiving nic
-                delay = calculatePropagationDelay(i->first);
-
-                sendDelayed(static_cast<cPacket*>(msg->dup()), delay, i->second);
-            }
-            // calculate delay (Propagation) to this receiving nic
-            delay = calculatePropagationDelay(i->first);
+    for (auto&& entry : gateList) {
+        const auto gate = entry.second;
+        const auto propagationDelay = calculatePropagationDelay(entry.first);
 
-            sendDelayed(msg, delay, i->second);
+        if (useSendDirect) {
+            for (int i = gate->getBaseId(); i < gate->getBaseId() + gate->size(); i++) {
+                sendDirect(msg->dup(), propagationDelay, msg->getDuration(), gate->getOwnerModule(), i);
+            }
```

Comment:
Thanks for the continued improvements. The code reads very well now.

If I could ask for one more cosmetic thing: rename the `i` variable to something clearer like `gateIndex` or so:
```suggestion
        if (useSendDirect) {
            for (int gateIndex = gate->getBaseId(); gateIndex < gate->getBaseId() + gate->size(); gateIndex++) {
                sendDirect(msg->dup(), propagationDelay, msg->getDuration(), gate->getOwnerModule(), gateIndex);
            }
```

And maybe some comment to the `delete msg;` at the end. Something like: Original message no longer needed, copies have been sent to all possible receivers.

Reply:
Done. Thanks. :)